src/mds/MDSRank.cc

```diff
@@ -2983,11 +2983,71 @@ bool MDSRankDispatcher::handle_command(
 
     damage_table.erase(id);
     return true;
+  } else if (prefix == "drop cache") {
+    command_drop_cache(ds);
+    return true;
   } else {
     return false;
   }
 }
 
+void MDSRank::command_drop_cache(stringstream *ds)
+{
+  if (not is_active()) {
+    dout(5) << __func__ << ": MDS not active, no-op" << dendl;
+    return;
+  }
+
+  if (not mdcache->trim(UINT64_MAX))
+    return;
+
+  server->recall_client_state(1, true);
+  mds_lock.Unlock();
+  if (not _wait_for_client_recall_state_acks())
+    dout(4) << "Timed out; didn't not receive "
+	    "CEPH_SESSION_RECALL_STATEMSG_ACK from all sessions." << dendl;
+
+  // Optionally, flush the journal -
+  Formatter *fmtr = new JSONFormatter(true);
+  command_flush_journal(fmtr);
+  // TODO: do we want output from `flush journal`?
+  fmtr->flush(*ds);
+  delete fmtr;
+
+  fmtr = new JSONFormatter(true);
+  int r = mdcache->cache_status(fmtr);
+  if (r != 0)
+   *ds << "Failed to get cache status: " << cpp_strerror(r);
+  fmtr->flush(*ds);
+}
+
+bool MDSRank::_wait_for_client_recall_state_acks()
+{
+  const unsigned int TIMEOUT = g_conf->get_val<uint64_t>("mds_drop_cache_ack_timeout");
+  set<Session*> unacked_sessions;
+  mono_time time_at_beg = mono_clock::now();
+  std::chrono::duration<double> time_passed = mono_clock::now() - time_at_beg;
+
+  sessionmap.get_client_session_set(unacked_sessions);
+
+  while (time_passed.count() < TIMEOUT) {
+    for (auto &session : unacked_sessions)
+      if (session->recall_acked)
+	unacked_sessions.erase(session);
+
+    if (unacked_sessions.empty())
+      return true;
+
+    time_passed = mono_clock::now() - time_at_beg;
+    if (TIMEOUT - time_passed.count() > 2)
+      sleep(2);
```

Comment:
This needs to be using waiter/gather continuations. @ukernel can you suggest the right approach.

Reply:
see Server::flush_client_sessions()

Reply:
I originally wrote this code to make MDS consume less CPU. Should it be gotten rid of? I am confused since discussion of waiter/gather continuations began from here.

Reply:
Done.

I was under the impression that `sleep` would lead to lesser number of iterations (when ACKs are not received spontaneously) allowing other threads/processes to use that CPU  time instead.